### ISSUE TITLE
[core] Noisy build logs

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#946]
Travis job logs are excessively noisy, see [this one](https://api.travis-ci.org/v3/job/348226443/log.txt) for example. Travis ends up truncating the log so that we can't see useful error messages. This is also a problem with local Maven builds, which is annoying :unamused: 

### maven-pmd-plugin

```
[INFO] --- maven-pmd-plugin:3.9.0:pmd (default) @ pmd-javascript ---
[WARNING] This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/pmd-6.0.1/pmd_userdocs_getting_started.html#incremental-analysis
...
[WARNING] This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/pmd-6.0.1/pmd_userdocs_getting_started.html#incremental-analysis
[WARNING] The RuleSet rulesets/java/basic.xml has been deprecated and will be removed in PMD 7.0.0
```

There are incremental analysis warnings and ruleset deprecation warnings, repeated dozens of times. Maybe #908 will help solve the latter

### ~Incremental analysis warnings~ (fixed by #954)
This is a problem both with the pmd plugin on travis, and with normal maven rule tests.  For example, for PerformanceRuleTests:
> [INFO] Running net.sourceforge.pmd.lang.java.rule.performance.PerformanceRulesTest
> Mar 02, 2018 1:26:07 PM net.sourceforge.pmd.cache.NoopAnalysisCache <init>
> WARNING: This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/latest/pmd_userdocs_getting_started.html#incremental-analysis
> ...

And that is repeated once per test. Maybe we should have an option `-no-cache` so that we can explicitly turn off the incremental analysis warning? It could be useful to users too

### ~ant.PMDTaskTest~ (fixed by #948) 

This test is exceptionnally noisy, printing a deprecated rule reference warning for every rule of pmd-java, e.g.
> Mar 02, 2018 1:25:02 PM net.sourceforge.pmd.RuleSetFactory parseRuleReferenceNode
> WARNING: Use Rule name category/java/errorprone.xml/SingleMethodSingleton instead of the deprecated Rule name rulesets/java/design.xml/SingleMethodSingleton. PMD 7.0.0 will remove support for this deprecated Rule name usage.

### pmd-xml 
I also noticed the following warning on `pmd-xml` during the dogfood task:

> [WARNING] Error during type resolution due to: java.lang.LinkageError: loader constraint violation in interface itable initialization: when resolving method "net.sourceforge.pmd.lang.xml.ast.AbstractDomNodeProxy.setUserData(Ljava/lang/String;Ljava/lang/Object;Lorg/w3c/dom/UserDataHandler;)Ljava/lang/Object;" the class loader (instance of net/sourceforge/pmd/util/ClasspathClassLoader) of the current class, net/sourceforge/pmd/lang/xml/ast/AbstractDomNodeProxy, and the class loader (instance of <bootloader>) for interface org/w3c/dom/Node have different Class objects for the type org/w3c/dom/UserDataHandler used in the signature

Again, it's repeated hundreds of times. It doesn't show on my local builds, but it's in Travis logs. This might have been caused by #901...

**Comment by adangel** [comment]
Additionally sourceforge is still not back to be usable atm - that's why the builds are failing, too. See https://twitter.com/sfnet_ops

About the maven-pmd-plugin warnings: yes, should be fixed with #908 . We could even now merge this as is (only minimum rules activated), since we anyway didn't fail the build till now with our own PMD violations...

Incremental analysis cache warning: It should display only once per PMD executing. So, this might be also the problem, that we execute PMD too often instead of reusing it? It could speed up the tests, too.
I can see the usefulness of a `-no-cache` option - basically explicitly disabling caching.

ant.PMDTaskTest: [pmdtasktest.xml](https://github.com/pmd/pmd/blob/master/pmd-java/src/test/resources/net/sourceforge/pmd/ant/xml/pmdtasktest.xml) really uses the old rulesets. It was simply missed for the 6.0.0 release and should be easy to fix.

pmd-xml: #901 is in 6.1.0, so we should be able to use 6.1.0 for our own dogfood task...

**Comment by adangel** [comment]
#948 addresses PMDTaskTest and CLITest, where we still used the old rulesets. Less warnings, but still not there yet.

**Comment by oowekyala** [comment]
> About the maven-pmd-plugin warnings: yes, should be fixed with #908 . We could even now merge this as is (only minimum rules activated), since we anyway didn't fail the build till now with our own PMD violations...

Yes, I think we should merge ASAP, else you'll have to monitor each incoming PR for trouble... We should probably wait for #943 though. We can then add rules one-by-one

> Incremental analysis cache warning: It should display only once per PMD executing. So, this might be also the problem, that we execute PMD too often instead of reusing it?

There is indeed [a new PMD for each test](https://github.com/pmd/pmd/blob/master/pmd-test/src/main/java/net/sourceforge/pmd/testframework/RuleTst.java#L261). Reusing the same could improve performance, or could not. My 2 cents about that: unless I'm overlooking something, reusing would have a big performance gain if we were executing several rules on the same piece of code, but test cases correspond to a single rule. If there's an easy way to isolate each test case from one another using the same PMD, then we should go for it, but it could easily become a nightmare trying to predict the side effects of every moving part. For now, running each test produces an [independent Report](https://github.com/pmd/pmd/blob/57e050f8c509ae873d40063a6e1783bb89380218/pmd-test/src/main/java/net/sourceforge/pmd/testframework/RuleTst.java#L139), which would be difficult to mimick with a unique PMD.

> pmd-xml: #901 is in 6.1.0, so we should be able to use 6.1.0 for our own dogfood task...

Nice find, so you think upgrading the pmd version of the plugin would fix it?

**Comment by oowekyala** [comment]
This can be closed I think

**Comment by adangel** [comment]
Just for reference, there are two bugs on the maven-pmd-plugin side related to this:

https://issues.apache.org/jira/browse/MPMD-257
https://issues.apache.org/jira/browse/MPMD-258

Both to be fixed with m-pmd-p 3.10.0
